iceoryx_posh/include/iceoryx_posh/experimental/popo/base_publisher.hpp

```diff
@@ -0,0 +1,125 @@
+
+namespace iox
+{
+namespace popo
+{
+using uid_t = UniquePortId;
+
+///
+/// @brief The PublisherInterface class defines the publisher interface used by the Sample class to make it generic.
+/// This allows any publisher specialization to be stored as a reference by the Sample class.
+///
+template <typename T>
+class PublisherInterface
+{
+  public:
+    virtual void publish(Sample<T> sample) noexcept = 0;
+
+  protected:
+    PublisherInterface() = default;
+};
+
+
+///
+/// @brief The BasePublisher class contains the common implementation for the different publisher specializations.
+///
+template <typename T, typename port_t = PublisherPortUser>
+class BasePublisher : public PublisherInterface<T>
+{
+  protected:
+    BasePublisher(const BasePublisher& other) = delete;
+    BasePublisher& operator=(const BasePublisher&) = delete;
+    BasePublisher(BasePublisher&& rhs) = default;
+    BasePublisher& operator=(BasePublisher&& rhs) = default;
+    ~BasePublisher() = default;
+
+    ///
+    /// @brief uid Get the UID of the publisher.
+    /// @return The publisher's UID.
+    ///
+    uid_t getUid() const noexcept;
+
+    ///
+    /// @brief loan Get a sample from loaned shared memory.
+    /// @param size The expected size of the sample.
+    /// @return An instance of the sample that resides in shared memory or an error if unable ot allocate memory to
+    /// laon.
+    /// @details The loaned sample is automatically released when it goes out of scope.
+    ///
+    cxx::expected<Sample<T>, AllocationError> loan(uint32_t size) noexcept;
+
+    ///
+    /// @brief publish Publish the given sample.
+    /// @param sample The sample to publish.
+    ///
+    void publish(Sample<T> sample) noexcept override;
+
+    ///
+    /// @brief previousSample Retrieve the previously loaned sample if it has not yet been claimed.
+    /// @return The previously loaned sample if retrieved.
+    ///
+    cxx::optional<Sample<T>> loanPreviousSample() noexcept;
+
+    ///
+    /// @brief offer Offer the service to be subscribed to.
+    ///
+    void offer() noexcept;
+
+    ///
+    /// @brief stopOffer Stop offering the service.
+    ///
+    void stopOffer() noexcept;
+
+    ///
+    /// @brief isOffered
+    /// @return True if service is currently being offered.
+    ///
+    bool isOffered() noexcept;
+
+    ///
+    /// @brief hasSubscribers
+    /// @return True if currently has subscribers to the service.
+    ///
+    bool hasSubscribers() noexcept;
+
+  protected:
+    BasePublisher(const capro::ServiceDescription& service);
+
+  private:
+    ///
+    /// @brief convertChunkHeaderToSample Helper function that wraps the payload of a ChunkHeader in an Sample.
+    /// @param header The chunk header describing the allocated memory chunk to use in the sample.
+    /// @return A sample that uses the ChunkHeader's payload as its memory allocation.
+    ///
+    Sample<T> convertChunkHeaderToSample(const mepoo::ChunkHeader* header) noexcept;
```

Comment:
Could you please turn it into a const pointer: `Sample<T> convertChunkHeaderToSample(const mepoo::ChunkHeader* const header) noexcept;`